openedx/custom/db/fields.py

```diff
@@ -0,0 +1,77 @@
+from django import VERSION
+from django.core import exceptions
+from django.db import models
+from django.utils.text import capfirst
+from multiselectfield import MultiSelectField
+from multiselectfield.db.fields import MSFList, add_metaclass
+from multiselectfield.utils import string_type
+
+from openedx.custom.forms.fields import MultiSelectFormFieldWithOthers
+
+
+class MultiSelectFieldWithOther(MultiSelectField):
+
+    def __init__(self, max_length, *args, **kwargs):
+        kwargs['max_length'] = max_length
+        if type(kwargs['choices']) == tuple:
+            kwargs['choices'] = kwargs.pop('choices') + (('other', 'Other'),)
+        super(MultiSelectFieldWithOther, self).__init__(*args, **kwargs)
+        self.error_messages.update({
+            'invalid_char': 'value %s contains invalid character `|`'
+        })
+
+    def get_prep_value(self, value):
+        return_value = ''
+        if value:
+            other_values = [val for val in value if not any(val in choice for choice in self.choices)]
+            value = [val for val in value if val not in other_values]
+            return_value = ('' if value is None else ",".join(map(str, value))) + ',|%s' % (
+                other_values.pop(0) if other_values else '',)
+        return return_value
+
+    def formfield(self, **kwargs):
+        defaults = {'required': not self.blank,
+                    'label': capfirst(self.verbose_name),
+                    'help_text': self.help_text,
+                    'choices': self.choices,
+                    'max_length': self.max_length,
+                    'max_choices': self.max_choices}
+        if self.has_default():
+            defaults['initial'] = self.get_default()
+        defaults.update(kwargs)
+        return MultiSelectFormFieldWithOthers(**defaults)
+
+    def validate(self, value, model_instance):
+        '''
+        this function is to validate the input values for multi select field, however we are implementing field with
+        support of other input filed we are disabling validations to let other input text pass to the database.
+        :param value: list of all selected choice with other text.
+        :param model_instance: current model instance for with it is saving data.
+        :return: nothing
+        '''
+        for opt_select in value:
+            if '|' in opt_select:
+                raise exceptions.ValidationError(self.error_messages['invalid_char'] % value)
```

Comment:
This validation is not enough for this model field, We are missing basic required validation. If field is required but user neither select any field nor enter text in other textfield it should raise validation error for required.. We are also not handling max_length validation here.. 
Form has required validation but model should also have it.
Please update this method with proper validations

Reply:
```
class MultiSelectField(models.CharField):
    """ Choice values can not contain commas. """

    def __init__(self, *args, **kwargs):
        self.min_choices = kwargs.pop('min_choices', None)
        self.max_choices = kwargs.pop('max_choices', None)
        super(MultiSelectField, self).__init__(*args, **kwargs)
        self.max_length = get_max_length(self.choices, self.max_length)
        self.validators[0] = MaxValueMultiFieldValidator(self.max_length)
        if self.min_choices is not None:
            self.validators.append(MinChoicesValidator(self.min_choices))
        if self.max_choices is not None:
            self.validators.append(MaxChoicesValidator(self.max_choices))
```

it is already handled in parent constructor with min and max choices.

Reply:
Those are different validations brother, Those are about minimum number of choices and max number of choices. 
 
I am talking about **required** and **max_length** validation.

Reply:
@zeshanarifios both are already implemented, as you have also verified on your machine.